### PR TITLE
Implement Parent tracking

### LINKED ISSUE #31: Connect store models with their parents
It would be nice to have something like `#parent` method, to get store model's parent object.

I just thought it might be useful. And I would definitely appreciate it in my project :)

It seems like it might be an additional step during object initialization. _(and in more places)_

I also think it is potentially related to https://github.com/DmitryTsepelev/store_model/issues/30, as it kind of "tightens" the connection between parent and store model.

### PR DESCRIPTION
**Pull request by blaze182**
G'day, trying to resolve #31 :)

This is still work in progress, needs some cleaning up, but parent is finally there, at least locally :)

Like `StoreModel::Types::*`, `ActiveModel::Attribute` initializer unfortunately doesn't have access to ActiveRecord/ActiveModel context as well, so I tried to introduce the patches to provide it during AttributeSet initialization and modification.

Hope the overall direction makes sense, open to any suggestions

### CODE REVIEW

File: lib/active_model/ext/attribute.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+module ActiveModel
+  class Attribute # :nodoc:
+    PARENT_TRACKING_ENABLED_TYPES = [
+      StoreModel::Types::ArrayType, StoreModel::Types::JsonType
+    ].freeze
+
+    private
+
+    attr_reader :context
```
[the remark this answers is not in the document]
Comment (blaze182) [reply lib/active_model/ext/attribute.rb]: Agree, makes sense!

File: lib/active_model/ext/attribute.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    def track_parent?
+      PARENT_TRACKING_ENABLED_TYPES.include?(type.class)
+    end
+
+    class FromDatabase < ActiveModel::Attribute # :nodoc:
```
[the remark this answers is not in the document]
Comment (blaze182) [reply lib/active_model/ext/attribute.rb]: Yeah that’s what I did initially, but it got ignored for some reason, so did a quick patch for now

File: lib/active_model/ext/attribute_set.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+require "active_model/attribute_set"
+
+ActiveModel::AttributeSet.prepend(Module.new do
```
[the remark this answers is not in the document]
Comment (blaze182) [reply lib/active_model/ext/attribute_set.rb]: Thanks for the tip, will do!

**Comment by blaze182** [comment]
Thanks for the feedback!
Regarding the override strategy, I was also thinking about just patching AR’s read_attribute, as it appeared to be the first place having both contexts in the caller chain. But yeah, as you said we need to cover at least three more cases, (1) attribute assignment, (2) same read/(3)write cases for plain `ActiveModel` instances (i.e. we can nest StoreModels). And same set for Array type, which in case of read/write override should have another condition I guess. Gotta add these to rspec suite btw :)

The motivation behind preserving parent reference and passing it top-down was attempt to eliminate possible edge-cases (i.e. reassignments, ActiveModel::Dirty tracking, attributes restoring and reloads etc.) by having the StoreModel initiated with parent straight away. As a bonus, we are leaving parent assignment logic encapsulated in Types :)
Anyway I can try looking into read/write override approach as well tomorrow.

File: lib/active_model/ext/attribute_set.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (blaze182) [reply lib/active_model/ext/attribute_set.rb]: done

**Comment by blaze182** [comment]
After digging deeper I found out that context approach isn't stable enough for attribute assignment case, so decided to proceed with read/write approach as you initially suggested. Looks like it's working now for AR objects for both json and array types :)
However still needs implementation for nested ActiveModels

**Comment by DmitryTsepelev** [comment]
Wow, that looks short! Great progress!

**Comment by blaze182** [comment]
Hi Dmitry, hope you're doing well!
Sorry was busy during the week, but finally got my proposed solution more less finalized during this weekend, so keen to hear your feedback :)

The only limitation I see now is direct StoreModel's Array mutation, it won't update a parent until one of the accessor methods gets called. As far as I can see it's achievable if we could introduce `StoreModel::Collection` instead of Array, but I guess it should be treated as another feature.

**Comment by DmitryTsepelev** [comment]
That looks awesome, thanks for the great work! I'll play with it a little bit on Thursday and come back with a feedback (if I bump into something). In the meantime, could you please add yourself into the changelog?

**Comment by blaze182** [comment]
Thank you, let me know how it goes :)
Added an entry for **master**

File: lib/store_model/ext/active_model/attributes.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+require "store_model/ext/parent_assignment"
+
+module StoreModel
+  # ActiveModel::Attributes patch with parent tracking support
+  module Attributes
+    include ParentAssignment
+
+    private
+
+    def attribute(*)
```
Comment (DmitryTsepelev) [inline lib/store_model/ext/active_model/attributes.rb]: Do you know the case when we need this patch? I've tried to remove it and all specs are passing, looks like this scenario is covered by `#_read_attribute` method

Comment (blaze182) [reply lib/store_model/ext/active_model/attributes.rb]: Good catch! We had a preemptive parent assignment on attribute write in specs :)
Now couple specs should fail accordingly

**Comment by DmitryTsepelev** [comment]
Merging, thanks again! 🎉

**Comment by DmitryTsepelev** [comment]
Hi @blaze182! Could you please send me your email and mailing address (we plan to send some goodies to CultOfMartians participants) to `dmitrytsepelev@evilmartians.com`?

**Comment by blaze182** [comment]
Woohoo, thanks @DmitryTsepelev! Just dropped you a line

**Pull request merged**